src/grpc/mod.rs

```diff
@@ -0,0 +1,197 @@
+            .success(result.clone())
+            .map_err(move |err| eprintln!("Failed to reply: {:?}", err))
+            .map(move |_| {
+                println!(
+                    "|**Warning**|: the WrapKey API is unimplemented! Responded with result {{ {:?} }}",
+                    result
+                )
+            });
+        ctx.spawn(wrap_output)
+    }
+}
+
+struct InputPayload {
+    kbc_name: String,
+    kbs_uri: String,
+    annotation: String,
+}
+
+fn parse_input(input_byte: Vec<u8>) -> Result<InputPayload> {
+    let input_string = String::from_utf8(input_byte)?;
+    println!("Request input {}", input_string);
+    let input: KeyProviderInput = serde_json::from_str::<KeyProviderInput>(&input_string[..])?;
+    let base64_annotation = input
+        .keyunwrapparams
+        .annotation
+        .ok_or_else(|| anyhow!("The annotation field in the input is Empty!"))?;
+    let vec_annotation = base64::decode(base64_annotation)?;
+    let jsonstring_annotation: &str = str::from_utf8(&vec_annotation)?;
+    let dc = input
+        .keyunwrapparams
+        .dc
+        .ok_or_else(|| anyhow!("The Dc field in the input is None!"))?;
+
+    /*
+     * AA expects the received DC parameter format is:
+     * "dc":{
+     *     "Parameters":{
+     *         "attestation_agent":["< KBC_NAME::KBS_URI (base64encode) >"]
+     *     }
+     * }
+     */
+    let parameters_list = dc.parameters.get("attestation_agent").ok_or(anyhow!(
+        "Invalid parameters: the request is not sent to attention agent!"
+    ))?;
+    let kbc_kbs_pair_byte = base64::decode(parameters_list[0].clone())?;
+    let kbc_kbs_pair = std::str::from_utf8(&kbc_kbs_pair_byte)?;
+    if let Some(index) = kbc_kbs_pair.find("::") {
+        let kbc_name: String = kbc_kbs_pair[..index].to_string();
+        let kbs_uri: String = kbc_kbs_pair[(index + 2)..].to_string();
+        println!("Get KBC_NAME:{}, KBS_URI:{}", kbc_name, kbs_uri);
+        return Ok(InputPayload {
+            kbc_name: kbc_name,
+            kbs_uri: kbs_uri,
+            annotation: jsonstring_annotation.to_string(),
+        });
+    } else {
+        return Err(anyhow!(
+            "Invalid parameters: invalid 'kbc_name::kbs_uri' pair format!"
+        ));
+    }
+}
+
+pub fn start_grpc_service(ip: IpAddr, port: u16) -> Result<()> {
+    let env = Arc::new(Environment::new(1));
+    let service = create_key_provider_service(KeyProvider);
+    let mut server = ServerBuilder::new(env)
+        .register_service(service)
+        .bind(&ip.to_string(), port)
+        .build()?;
+
+    server.start();
+    println!("Keyprovider gRPC service started!");
+
+    let (tx, rx) = oneshot::channel();
+    thread::spawn(move || {
+        println!("Press ENTER to exit...");
+        match io::stdin().read(&mut [0]) {
+            Ok(_) => (),
+            Err(_) => panic!("Exit command unexpect!"),
+        };
+        tx.send(())
+    });
+    let _ = block_on(rx);
```

Comment:
Obviously no one is going to press a key when the attestation agent either runs as a guest service or is spawned from the kata-agent.
I think here you want to handle signals (SIGINT, SIGTERM) as this is typically what this program will receive when we're shutting it down.

So using something like the`signal-hook` crate would allow for registering a `block_on(server.shutdown())` on  either SIGINT or SIGTERM.